feat: Introduce `max_cycles_limit` of a Godwoken block

zeroqn opened the pull request:
## Summary

This pr adds a new cycles limit `max_cycles_limit` to mem block. If mem pool's available cycles is not enough, transactions will be packaged in next block. If a transaction's cycles exceeded `max_cycles_limit`, it will be discarded.

## Motivation

A dynamic configurable way to manage number of block transaction.

## Specification

Split transaction's cycles into two parts: `execution cycles` and `virtual cycles`. Only syscalls listed in `SyscallCyclesConfig` contribute to transaction's `virtual cycles`.

### SyscallCyclesConfig

```rust
pub struct SyscallCyclesConfig {
    pub sys_store_cycles: u64,
    pub sys_load_cycles: u64,
    pub sys_create_cycles: u64,
    pub sys_load_account_script_cycles: u64,
    pub sys_store_data_cycles: u64,
    pub sys_load_data_cycles: u64,
    pub sys_get_block_hash_cycles: u64,
    pub sys_recover_account_cycles: u64,
    pub sys_log_cycles: u64,
}
```

## Different between `L2TX_MAX_CYCLES` and block `max_cycles_limit`

| L2TX_MAX_CYCLES  | Block max cycles limit |
| :---:    | :---:    |
| execution cycles  | execution cycles + virtual cycles  |
| failed transaction | discard | 


## Implementation

- new(config): add `max_cycles_limit` and `syscall_cycles: SyscallCyclesConfig` to `MemBlockConfig`
- feat(generator): `TransactionError::ExceededMaxBlockCycles` and `TransactionError::InsufficientPoolCycles`
- feat(rpc-server): stop package tx into mem block when mem pool reaches `max_cycles_limit`
- feat(rpc-server): discard tx if it's cycles limit exceed block `max_cycles_limit`
- feat(rpc-server): new `TransactionError::ExceededMaxBlockCycles` for `gw_execute_l2transacton` and `gw_execute_raw_l2transaction`

zeroqn committed:
feat: max block cycles limit

zeroqn committed:
fix(mem-pool): use L2TX_MAX_CYCLES when block max cycles is smaller

zeroqn committed:
fix(rpc-server): restore mem block txs reach block max cycles limit

zeroqn committed:
fix(mem-pool): use remained cycles as block max cycles limit

zeroqn committed:
test(rpc-server): block max cycles limit for submit_l2transaction

zeroqn committed:
fix(generator): subtract used syscall cycles twice

zeroqn committed:
test(rpc-server): block max cycles limit for execute api

zeroqn committed:
refactor(generator): rename TransactionError::CyclesLimitReached

Rename to TransactionError::BlockCyclesLimitReached

zeroqn committed:
fix(generator): tx exceed block max cycles also returns BlockCyclesLimitReached

zeroqn committed:
refactor(rpc-server): log block cycles limit reached for tx

zeroqn committed:
fix(rpc-server): log error on tx cycles limit exceed block max cycles

Use log::info instead

zeroqn commented:
/itest
tests: 8cd2fda0e8ade88fc99ca3996435655fbc135bb1
web3: ca5a7fdfb83e757d56b3c81641af2199386828a0
scripts: master

zeroqn commented:
/itest
tests: 8cd2fda0e8ade88fc99ca3996435655fbc135bb1
web3: ca5a7fdfb83e757d56b3c81641af2199386828a0

zeroqn commented:
/itest
web3: a073759352a7f9d80470279d8660dfef18340486

zeroqn commented:
/itest
tests: 73ca1bf2496fc3a08956105f50be0f21276374f1

zeroqn committed:
fix(config): u64::MAX deserialize failure for max_cycles_limit

zeroqn commented:
/itest
web3: main

blckngm commented on crates/config/src/config.rs:
```diff
@@ -432,3 +477,33 @@ impl Default for ContractLogConfig {
+}
+
+impl SyscallCyclesConfig {
+    pub fn all_zero() -> Self {
```
Why not derive `Default`?

zeroqn replied:
To make explicit

blckngm commented on crates/rpc-server/src/registry.rs:
```diff
@@ -607,16 +635,44 @@ impl RequestSubmitter {
                 // removed from the in queue request map after pushed to mem
                 // pool.
-                for (entry, _handle) in items {
+                for (entry, handle) in items {
```
The comment above is outdated now. Let's remove it.

zeroqn replied:
Ok

zeroqn committed:
fix(registry): remove outdated comment about queue `_handle`

jjyr commented on crates/generator/src/generator.rs:
```diff
@@ -92,6 +92,73 @@ impl From<WithdrawalCellError> for Error {
+#[derive(Debug, Clone)]
+pub struct CyclesPool {
+    limit: u64,
+    cycles: u64,
```
`remained_cycles` or `available_cycles` wound be easy to understand

jjyr commented on crates/generator/src/generator.rs:
```diff
@@ -92,6 +92,73 @@ impl From<WithdrawalCellError> for Error {
+    limit: u64,
+    cycles: u64,
+    syscall_config: SyscallCyclesConfig,
+    run_out: bool,
```
Can we remove this field, we can use `remained_cycles == 0` to detect run out

jjyr commented on crates/generator/src/generator.rs:
```diff
@@ -92,6 +92,73 @@ impl From<WithdrawalCellError> for Error {
+        &self.syscall_config
+    }
+
+    pub fn sub_cycles(&mut self, cycles: u64) {
```
`consume_cycles` since we do a mutate here.

jjyr commented on crates/generator/src/generator.rs:
```diff
@@ -92,6 +92,73 @@ impl From<WithdrawalCellError> for Error {
+    raw_tx: &'a RawL2Transaction,
+    max_cycles: u64,
+    backend: Backend,
+    cycles_pool: &'a mut CyclesPool,
```
We can name the struct and the fields `cycle_meter: CycleMeter`

jjyr commented on crates/generator/src/generator.rs:
```diff
@@ -123,22 +190,28 @@ impl Generator {
-        let used_cycles;
+        let mut used_cycles;
         let exit_code;
+        let cycles_pool_bak = cycles_pool.clone();
```
Use the `origin_cycle_meter` or `cycle_meter_before_execute` to make the code easy to read.

jjyr commented on crates/generator/src/generator.rs:
```diff
@@ -179,16 +253,36 @@ impl Generator {
+            // NOTE: already subtract syscall cycles from cycles pool during machine running to
+            // interrupt execution eariler.
+            let used_syscall_cycles = cycles_pool_bak.cycles() - cycles_pool.cycles();
+            cycles_pool.sub_cycles(used_cycles.saturating_sub(used_syscall_cycles));
```
Why minus the syscall cycles? By my understand the mem-pool should reject a tx if it consumes too many syscall cycles, it that right?

zeroqn replied:
Machine cycles include syscall cycles. And we already minus syscall cycles in syscalls implementation [here](
https://github.com/nervosnetwork/godwoken/pull/767/files#diff-16f67f4822618c8e4616f5d07b43084312bdba514bf01f1c696cd16a86a6d496R152)

jjyr commented on crates/generator/src/syscalls/mod.rs:
```diff
@@ -602,4 +619,19 @@ impl<'a, S: State, C: ChainView> L2Syscalls<'a, S, C> {
+            SYS_GET_BLOCK_HASH => cycles_config.sys_get_block_hash_cycles,
+            SYS_RECOVER_ACCOUNT => cycles_config.sys_recover_account_cycles,
+            SYS_LOG => cycles_config.sys_log_cycles,
+            _ => 0,
```
I think we should panic here, return `0` may ignoring unintended behaviours

zeroqn replied:
But we only include partial syscalls in `SyscallCyclesConfig`.
0 is intended default value for syscalls not in config.

zeroqn committed:
fix(generator): modify cycles config affect history tx run result

zeroqn committed:
feat(generator): interrupt execution eariler if exceeded block max cycles

zeroqn committed:
refactor(types): remove unused used_cycles in RunResult

replace by cycles.execution

jjyr commented on crates/generator/src/error.rs:
```diff
@@ -159,6 +159,10 @@ pub enum TransactionError {
+    #[error("Block cycles limit reached: cycles {cycles:?}, limit {limit}")]
+    BlockCyclesLimitReached { cycles: RunResultCycles, limit: u64 },
+    #[error("Exceeded block max cycles limit: cycles {cycles:?}, limit {limit}")]
+    ExceededBlockMaxCycles { cycles: RunResultCycles, limit: u64 },
```
What's the difference of these two errors?

jjyr replied:
I suggest we rename the two errors to make some distinguish

* `BlockCyclesLimitReached` -> `ExceededMaxBlockCycles` - the tx consumes cycles beyond the block maximum cycles, we cannot package the tx at all.
* `ExceededBlockMaxCycles` -> `InsufficientPoolCycle` - we have not enough pool cycles for the tx, we can repackage the tx in the next block.

jjyr commented on crates/generator/src/generator.rs:
```diff
@@ -92,6 +92,60 @@ impl From<WithdrawalCellError> for Error {
+        }
+    }
+
+    pub fn none<'a>() -> Option<&'a mut CyclesPool> {
```
If we use the `Option<_>` type, I think we can just remove this and pass `None` to the function.

jjyr commented on crates/generator/src/generator.rs:
```diff
@@ -179,16 +240,45 @@ impl Generator {
+
+            // Subtract tx execution cycles.
+            if let Some(cycles_pool) = &mut cycles_pool {
+                if cycles_pool.checked_sub_cycles(execution_cycles).is_none() {
```
We should use the name `consume_cycles`, it is hard to see this function had side-effect when read the code.

zeroqn committed:
refactor(generator): base on review

rename BlockCyclesLimitReached to InsufficientPoolCycles
remove CyclesPool::none, use None instead
rename check_sub_cycles to consume_cycles

zeroqn committed:
test: only virtual cycles exceeded block max cycles

Execution cycles < block max cycles
Execution cycles < L2TX_MAX_CYCLES
Execution + virtual cycles > block max cycles

zeroqn committed:
fix(rpc-server): log message for withdrawal cycles > max block cycles

zeroqn committed:
chore(ci): set log level to debug for cargo test

zeroqn committed:
fix(tests): submit_withdrawal_request random failure

Withdrawal request is pushed to mem pool immediately
https://github.com/nervosnetwork/godwoken/runs/7533962286?check_suite_focus=true#step:9:741

In this case, we should produce new block with withdrawal request to
update state.

zeroqn committed:
fix(generator): add 'drop' statement on ExceededMaxBlockCycles Error

zeroqn committed:
refactor(generator): remove redundant lifetime parameter
